cpp/src/arrow/filesystem/gcsfs_test.cc

```diff
@@ -414,6 +414,34 @@ TEST_F(GcsIntegrationTest, CreateDirRecursiveBucketAndFolder) {
   arrow::fs::AssertFileInfo(fs.get(), "new-bucket/", FileType::Directory);
 }
 
+TEST_F(GcsIntegrationTest, DeleteDirSuccess) {
+  auto fs = internal::MakeGcsFileSystemForTest(TestGcsOptions());
+  const char* const kTestFolders[] = {
+      "a/", "a/0/", "a/0/0/", "a/1/", "a/2/",
+  };
+  for (auto const* f : kTestFolders) {
+    const auto folder = PreexistingBucketPath() + f;
+    ASSERT_OK(fs->CreateDir(folder, true));
+    for (int i = 0; i != 64; ++i) {
+      const auto filename = folder + "test-file-" + std::to_string(i);
+      ASSERT_OK_AND_ASSIGN(auto w, fs->OpenOutputStream(filename, {}));
+      ASSERT_OK(w->Write(filename.data(), filename.size()));
+      ASSERT_OK(w->Close());
+    }
+  }
+
+  ASSERT_OK(fs->DeleteDir(PreexistingBucketPath() + kTestFolders[0]));
+
+  for (auto const* f : kTestFolders) {
+    const auto folder = PreexistingBucketPath() + f;
+    arrow::fs::AssertFileInfo(fs.get(), folder, FileType::NotFound);
+    for (int i = 0; i != 64; ++i) {
+      const auto filename = folder + "test-file-" + std::to_string(i);
+      arrow::fs::AssertFileInfo(fs.get(), filename, FileType::NotFound);
+    }
+  }
+}
```

Comment:
Should you test that other paths in the bucket still exist?

Reply:
Done.